File: logs/src/main/java/com/stone/log/Logs.java
package com.stone.log;

import android.text.TextUtils;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created By: sqq
 * Created Time: 9/6/18 4:15 PM.
 * <p>
 * 基于android.util.Log日志输出的封装工具类
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public class Logs {

    public static final int VERBOSE = 1;
    public static final int DEBUG = 2;
    public static final int INFO = 3;
    public static final int WARN = 4;
    public static final int ERROR = 5;

    /**
     * json格式化显示的缩进字符数
     */
    private final static int JSON_INDENT = 4;

    /**
     * 应用名标签（用以本应用Log打印的默认TAG）
     */
    private static String TAG = "logs";

    /**
     * log输出等级
     */
    private static int level = DEBUG;

    /**
     * 行分隔符
     */
    private static final String LINE_SEPARATOR = System.getProperty("line.separator");

    /**
     * 单条log最大输出限制，超出会自动分段处理；
     * <p>
     * android.util.Log中默认最大输出长度会稍微大于这个值
     */
    private static final int MAX_SINGLE_LOG_LENGTH = 3 * 1024;

    private Logs() {
        /* cannot be instantiated */
        throw new UnsupportedOperationException("cannot be instantiated");
    }

    /**
     * 初始化Logs
     *
     * @param level      控制log的输出等级
     * @param defaultTag 默认的全局log TAG
     */
    public static void init(int level, String defaultTag) {
        Logs.level = level;
        Logs.TAG = defaultTag;
    }

    public static void i(Object str) {
        i("", str);
    }

    public static void d(Object str) {
        d("", str);
    }

    public static void v(Object str) {
        v("", str);
    }

    public static void w(Object str) {
        w("", str);
    }

    public static void e(Object str) {
        e("", str);
    }

    public static void i(String tag, Object str) {
        if (level <= INFO) {
            String name = getFunctionName();
            tag = getTag(tag);
            log(INFO, tag, name + " - " + str);
        }

    }

    public static void d(String tag, Object str) {
        if (level <= DEBUG) {
            String name = getFunctionName();
            tag = getTag(tag);
            log(DEBUG, tag, name + " - " + str);
        }
    }

    private static String getTag(String tag) {
        return tag + TAG;
    }

    public static void v(String tag, Object str) {
        if (level <= VERBOSE) {
            String name = getFunctionName();
            tag = getTag(tag);
            log(VERBOSE, tag, name + " - " + str);
        }
    }

    public static void w(String tag, Object str) {
        if (level <= WARN) {
            String name = getFunctionName();
            tag = getTag(tag);
            log(WARN, tag, name + " - " + str);
        }
    }

    public static void e(String tag, Object str) {
        if (level <= ERROR) {
            String name = getFunctionName();
            tag = getTag(tag);
            log(ERROR, tag, name + " - " + str);
        }
    }

    public static void e(Exception ex) {
        if (level <= ERROR) {
            Log.e(TAG, "error", ex);
        }
    }

    public static void e(String tag, String msg, Throwable tr) {
        if (level <= ERROR) {
            String name = getFunctionName();
            tag = getTag(tag);
            Log.e(tag, name + " - " + msg, tr);
        }
    }

    public static void json(Object json) {
        json(TAG, json, null);
    }

    public static void json(Object json, String url) {
        json(TAG, json, url);
    }


    private static void log(int level, String tag, Object msg) {
        if (msg == null) showLog(level, tag, "null");
        else {
            int length = msg.toString().length();
            int start = 0;
            int end = MAX_SINGLE_LOG_LENGTH;
            int count = length / MAX_SINGLE_LOG_LENGTH;
            for (int i = 0; i <= count; i++) {
                if (length > end) {
                    showLog(level, tag, msg.toString().substring(start, end));
                    start = end;
                    end += MAX_SINGLE_LOG_LENGTH;
                } else {
                    showLog(level, tag, msg.toString().substring(start, length));
                    break;
                }
            }
        }

    }

    private static void showLog(int level, String tag, Object msg) {
        switch (level) {
            case VERBOSE:
                Log.v(tag, msg.toString());
                break;
            case DEBUG:
                Log.d(tag, msg.toString());
                break;
            case INFO:
                Log.i(tag, msg.toString());
                break;
            case WARN:
                Log.w(tag, msg.toString());
                break;
            case ERROR:
                Log.e(tag, msg.toString());
                break;
        }
    }

    /**
     * 格式化输出json
     */
    public static void json(String tag, Object json, String url) {
        if (level <= DEBUG) {
            String name = getFunctionName();
            Log.d(tag, "╔═══════════════════════════════════════════════════════════════════════════════════════");
            if (!TextUtils.isEmpty(url))
                Log.i(tag, "║ onResponseSuccess URL：" + url);
            printJson(tag, name, json.toString());
            Log.d(tag, "╚═══════════════════════════════════════════════════════════════════════════════════════");
        }
    }

    private static void printJson(String tag, String name, String json) {
        String message;
        try {
            if (json.startsWith("{")) {
                JSONObject jsonObject = new JSONObject(json);
                message = jsonObject.toString(JSON_INDENT);
            } else if (json.startsWith("[")) {
                JSONArray jsonArray = new JSONArray(json);
                message = jsonArray.toString(JSON_INDENT);
            } else {
                message = json;
            }
        } catch (JSONException e) {
            message = json;
        }
        message = name + LINE_SEPARATOR + message;
        String[] lines = message.split(LINE_SEPARATOR);
        for (String line : lines) {
            Log.d(tag, "║ " + line);
        }
    }


    private static String getFunctionName() {
        return getFunctionName(false);
    }

    /**
     * 获取当前线程、类、方法、行
     */
    @SuppressWarnings("SameParameterValue")
    private static String getFunctionName(boolean isThreadBreak) {
        StackTraceElement[] sts = Thread.currentThread().getStackTrace();
        if (sts == null) return "";
        for (StackTraceElement st : sts) {
            if (st.isNativeMethod()) {
                continue;
            }
            if (st.getClassName().equals(Thread.class.getName())) {
                continue;
            }
            if (st.getClassName().equals(Logs.class.getName()) || st.getClassName().equals(Logs.class.getName() + "$Companion")) {
                continue;
            }
            if (isThreadBreak) {
                return "[ " + Thread.currentThread().getName() + ": \n("
                        + st.getFileName() + ":" + st.getLineNumber() + ") "
                        + st.getMethodName() + " ]";
            } else {
                return "[ " + Thread.currentThread().getName() + ": ("
                        + st.getFileName() + ":" + st.getLineNumber() + ") "
                        + st.getMethodName() + " ]";
            }
        }
        return "";
    }
}

File: logs/src/main/java/com/stone/log/Logs.kt
package com.stone.log

import android.text.TextUtils
import android.util.Log


import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * Created by sqq on 17/6/1
 */
class Logs private constructor() {


    init {
        /* cannot be instantiated */
        throw UnsupportedOperationException("cannot be instantiated")
    }

    companion object {
        const val VERBOSE = 1
        const val DEBUG = 2
        const val INFO = 3
        const val WARN = 4
        const val ERROR = 5
        /**
         * 是否输出log
         */
        private var level = DEBUG
        /**
         * json格式化显示的缩进字符数
         */
        private const val JSON_INDENT = 4
        /**
         * 应用Log标签（用以本应用Log打印的默认TAG）
         */
        private var TAG = ""
        /**
         * 行分隔符
         */
        private val LINE_SEPARATOR = System.getProperty("line.separator")


        private const val MAX_SINGLE_LOG_LENGTH = 3 * 1024//单条log系统默认最大输出限制，超出会自动截掉

        /**
         * 初始化Logs
         * @param level 控制log的输出等级
         * @param defaultTag 默认的全局log TAG
         */
        fun init(level: Int = DEBUG, defaultTag: String = "logs") {
            this.level = level
            this.TAG = defaultTag
        }

        fun i(str: Any) {
            i("", str)
        }

        fun d(str: Any) {
            d("", str)
        }

        fun v(str: Any) {
            v("", str)
        }

        fun w(str: Any) {
            w("", str)
        }

        fun e(str: Any) {
            e("", str)
        }

        fun i(tag: String, str: Any) {
            if (level <= INFO) {
                val name = functionName
                val t = getTag(tag)
                if (name != null) {
                    log(INFO, t, "$name - $str")
                } else {
                    log(INFO, t, str.toString())
                }
            }
        }

        fun d(tag: String, str: Any) {
            if (level <= DEBUG) {
                val name = functionName
                val t = getTag(tag)
                if (name != null) {
                    log(DEBUG, t, "$name - $str")
                } else {
                    log(DEBUG, t, str.toString())
                }
            }
        }

        private fun getTag(tag: String): String {
            return tag + TAG
        }

        fun v(tag: String, str: Any) {
            if (level <= VERBOSE) {
                val name = functionName
                val t = getTag(tag)
                if (name != null) {
                    log(VERBOSE, t, "$name - $str")
                } else {
                    log(VERBOSE, t, str.toString())
                }
            }
        }

        fun w(tag: String, str: Any) {
            if (level <= WARN) {
                val name = functionName
                val t = getTag(tag)
                if (name != null) {
                    log(WARN, t, "$name - $str")
                } else {
                    log(WARN, t, str.toString())
                }
            }
        }

        fun e(tag: String, str: Any) {
            if (level <= ERROR) {
                val name = functionName
                val t = getTag(tag)
                if (name != null) {
                    log(ERROR, t, "$name - $str")
                } else {
                    log(ERROR, t, str.toString())
                }
            }
        }

        fun e(ex: Exception) {
            if (level <= ERROR) {
                Log.e(TAG, "error", ex)
            }
        }

        fun e(log: String, tr: Throwable) {
            if (level <= ERROR) {
                val line = functionName
                Log.e(TAG, "{Thread:" + Thread.currentThread().name + "}"
                        + "[" + line + ":] " + log + "\n", tr)
            }
        }

        fun json(json: Any) {
            json(TAG, json, null)
        }

        fun json(json: Any, url: String) {
            json(TAG, json, url)
        }

        private fun log(level: Int, tag: String, msg: Any?) {
            if (msg != null) {
                val length = msg.toString().length
                if (length > MAX_SINGLE_LOG_LENGTH) {
                    var start = 0
                    var end = MAX_SINGLE_LOG_LENGTH
                    val count = length / MAX_SINGLE_LOG_LENGTH
                    for (i in 0..count) {
                        if (length > end) {
                            showLog(level, tag, msg.toString().substring(start, end))
                            start = end
                            end += MAX_SINGLE_LOG_LENGTH
                        } else {
                            showLog(level, tag, msg.toString().substring(start, length))
                            break
                        }
                    }
                } else {
                    showLog(level, tag, msg)
                }
            }
        }

        private fun showLog(level: Int, tag: String, msg: Any) {
            when (level) {
                VERBOSE -> Log.v(tag, msg.toString())
                DEBUG -> Log.d(tag, msg.toString())
                INFO -> Log.i(tag, msg.toString())
                WARN -> Log.w(tag, msg.toString())
                ERROR -> Log.e(tag, msg.toString())
                else -> Log.d(tag, msg.toString())
            }
        }

        /**
         * 格式化输出json
         *
         * @param tag
         * @param json
         */
        fun json(tag: String, json: Any, url: String?) {
            if (level <= DEBUG) {
                var name = functionName
                if (name == null) name = ""
                Log.d(tag, "╔═══════════════════════════════════════════════════════════════════════════════════════")
                if (!TextUtils.isEmpty(url))
                    Log.w(tag, "║ onResponseSuccess URL：" + url!!)
                printJson(tag, name, json.toString())
                Log.d(tag, "╚═══════════════════════════════════════════════════════════════════════════════════════")
            }
        }

        /**
         * 格式化json
         */
        private fun printJson(tag: String, name: String, json: String) {
            var message: String
            message = try {
                when {
                    json.startsWith("{") -> {
                        val jsonObject = JSONObject(json)
                        jsonObject.toString(JSON_INDENT)
                    }
                    json.startsWith("[") -> {
                        val jsonArray = JSONArray(json)
                        jsonArray.toString(JSON_INDENT)
                    }
                    else -> json
                }
            } catch (e: JSONException) {
                json
            }

            message = name + LINE_SEPARATOR + message
            val lines = message.split(LINE_SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (line in lines) {
                log(DEBUG, tag, "║ $line")
            }
        }


        private val functionName: String?
            get() = getFunctionName(false)

        /**
         * 获取当前线程、类、方法、行
         */
        private fun getFunctionName(isThreadBreak: Boolean): String? {
            val sts = Thread.currentThread().stackTrace ?: return null
            for (st in sts) {
                if (st.isNativeMethod) {
                    continue
                }
                if (st.className == Thread::class.java.name) {//java.lang.Thread
                    continue
                }
                if (st.className == Logs::class.java.name || st.className == "${Logs::class.java.name}\$Companion") {
                    continue
                }
                return if (isThreadBreak) {
                    ("[ " + Thread.currentThread().name + ": \n("
                            + st.fileName + ":" + st.lineNumber + ") "
                            + st.methodName + " ]")
                    // fileName=Logs.kt
                } else {
                    ("[ " + Thread.currentThread().name + ": ("
                            + st.fileName + ":" + st.lineNumber + ") "
                            + st.methodName + " ]")
                }
            }
            return null
        }
    }
}
